client/src/controllers/FormsetController.ts

```diff
@@ -0,0 +1,285 @@
+  declare totalValue: number;
+
+  initialize() {
+    this.totalValue = parseInt(this.totalFormsInputTarget.value, 10);
+    this.minValue = parseInt(this.minFormsInputTarget.value, 10);
+    this.maxValue = parseInt(this.maxFormsInputTarget.value, 10);
+  }
+
+  connect() {
+    this.dispatch('ready', {
+      cancelable: false,
+      detail: {
+        minValue: this.minValue,
+        maxValue: this.maxValue,
+        totalValue: this.totalValue,
+      },
+    });
+  }
+
+  /**
+   * Add a new child form from the template content.
+   */
+  add() {
+    if (this.childTargets.length >= this.maxValue) return;
+
+    if (
+      this.dispatch('adding', {
+        cancelable: true,
+        detail: { formIndex: this.totalValue },
+      }).defaultPrevented
+    ) {
+      return;
+    }
+
+    this.formsTarget.append(this.newChild);
+  }
+
+  /**
+   * Find the event's target's closest child target and remove it by
+   * removing the 'child' target and adding a 'child-removed' target.
+   */
+  delete(event: Event) {
+    const target = this.childTargets.find((child) =>
+      child.contains(event.target as Node),
+    );
+
+    if (!target) {
+      throw new Error(
+        `Could not find child form target for event target: ${event.target}.`,
+      );
+    }
+
+    if (this.childTargets.length <= this.minValue) return;
+
+    if (
+      this.dispatch('removing', { target, cancelable: true }).defaultPrevented
+    ) {
+      return;
+    }
+
+    const targetAttrName = `data-${this.identifier}-target`;
+
+    target.setAttribute(
+      targetAttrName,
+      (target.getAttribute(targetAttrName)?.split(' ') ?? [])
+        .filter((name) => name !== 'child')
+        .concat(['deleted'])
+        .join(' '),
+    );
+  }
+
+  /**
+   * When a new child is added, update the total count and dispatch an added event.
+   */
+  childTargetConnected(target: HTMLElement) {
+    const totalFormsCount =
+      this.childTargets.length + this.deletedTargets.length;
+    if (totalFormsCount === this.totalValue) return;
+    this.totalValue = totalFormsCount;
+    this.dispatch('added', {
+      target,
+      cancelable: false,
+      detail: { formIndex: totalFormsCount - 1 },
+    });
+  }
+
+  /**
+   * When removed, add the class and update the total count.
+   * Also update the DELETE input for this form.
+   */
+  childTargetDisconnected(target: HTMLElement) {
+    const deletedClasses = this.deletedClasses;
+
+    afterTransition(
+      target,
+      // If there are no classes to add, we can skip the delay before hiding.
+      deletedClasses.length ? {} : { maxDelay: 0 },
+    )(target.classList.add(...deletedClasses)).then(() => {
+      target.setAttribute('hidden', '');
+      this.dispatch('removed', { target, cancelable: false });
+    });
+
+    const deleteInput = this.deleteInputTargets.find((input) =>
+      target.contains(input),
+    );
+
+    if (!deleteInput) {
```

Comment:
Note. I could not easily write a unit test for this error case due to Stimulus not catching these errors. 

https://github.com/hotwired/stimulus/issues/711

I have manually tested this scenario though and it does correctly throw in the console without breaking the whole Controller behaviour.